fuzz/fuzz_types.ml

```
@@ -0,0 +1,440 @@
+  | TResult (ta, te), VResult v -> result_map (unwrap ta) (unwrap te) v
+  | TRecord _, VRecord v -> v
+  | TVariant _, VVariant v -> v
+  | TEnum _, VEnum v -> v
+  | _ ->
+      failwith
+      @@ fmt "Tried to unwrap %s while expecting type %s." (show_any_value w)
+           (show_any_type (AT t))
+
+(** Check whether [l] has unique elements using comparison function [cmp]. *)
+let is_unique_list (type a) ?(cmp : a -> a -> int = compare) =
+  let module S = Set.Make (struct
+    type t = a
+
+    let compare = cmp
+  end) in
+  let rec aux set = function
+    | [] -> true
+    | x :: xs when S.find_opt x set = None -> aux (S.add x set) xs
+    | _ -> false
+  in
+  aux S.empty
+
+exception Utf8_failure
+
+(** Check whether the string only contains valid UTF-8 characters. *)
+let is_valid_utf8 str =
+  try
+    Uutf.String.fold_utf_8
+      (fun _ _ -> function `Malformed _ -> raise Utf8_failure | _ -> ())
+      () str;
+    true
+  with Utf8_failure -> false
+
+(** Ensure that a [string list] is suited for Irmin field names. *)
+let guard_strings l =
+  guard
+    ( List.length l <= 5
+    && List.for_all is_valid_utf8 l
+    && is_unique_list ~cmp:String.compare l );
+  l
+
+(** Ensure that a [(string * 'a) list] is suited for Irmin field names. *)
+let guard_keys l =
+  List.map fst l |> guard_strings |> ignore;
+  l
+
+(** Generate a dynamic type recursively. *)
+let any_type_gen =
+  fix (fun at_gen ->
+      let field_gen = pair string at_gen in
+      let case_type_gen =
+        choose
+          [ const (ACT Case0); map [ at_gen ] (fun (AT at) -> ACT (Case1 at)) ]
+      in
+      let case_gen = pair string case_type_gen in
+      let fields_gen = map [ list1 field_gen ] guard_keys in
+      let cases_gen = map [ list1 case_gen ] guard_keys in
+      let tags_gen = map [ list1 string ] guard_strings in
+      choose
+        [
+          const (AT TUnit);
+          const (AT TBool);
+          const (AT TChar);
+          (* const (AT TInt);
```

you forgot to uncomment these

---

There is a bit of a problem here, actually. In the fuzzer, we're testing both the binary and the JSON encoder/decoder. But, when encoding numerical values, the JSON encoder is forced to convert everything to floating-point because of the way the `Jsonm` API is designed:

```ocaml
let float e f = lexeme e (`Float f)
let int e i = float e (float_of_int i)
let int32 e i = float e (Int32.to_float i)
let int64 e i = float e (Int64.to_float i)
```

This leads to a loss of precision which raises a lot of "unwanted" fuzzing errors (as the decoded value differs from the original one). I tried to ignore them using a custom equality function which would tolerate numerical errors up to a certain threshold, but that led to a lot more boilerplate code. So I'm wondering if it would not be simpler to just avoid using numerical values when fuzzing the JSON encoder?

---

It makes sense to not use these types when fuzzing json. Can you add a comment about this?

---

I disagree, and think this is a deficiency of the `Jsonm` API; the choice of `float` as an internal representation is a fairly arbitrary one.

If we're going to do things like e.g. storing distributions taken from Index benchmark results in Irmin stores, I think it's suspicious to silently drop precision of numerical values. It's not related to this PR, so perhaps just warrants a separate issue.